zarr/storage.py

```diff
@@ -1877,6 +1878,197 @@ def __delitem__(self, key):
             self._invalidate_value(key)
 
 
+class SQLiteStore(MutableMapping):
+    """Storage class using SQLite.
+
+    Parameters
+    ----------
+    path : string
+        Location of database file.
+    **kwargs
+        Keyword arguments passed through to the `sqlite3.connect` function.
+
+    Examples
+    --------
+    Store a single array::
+
+        >>> import zarr
+        >>> store = zarr.SQLiteStore('data/array.sqldb')
+        >>> z = zarr.zeros((10, 10), chunks=(5, 5), store=store, overwrite=True)
+        >>> z[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+
+    Store a group::
+
+        >>> store = zarr.SQLiteStore('data/group.sqldb')
+        >>> root = zarr.group(store=store, overwrite=True)
+        >>> foo = root.create_group('foo')
+        >>> bar = foo.zeros('bar', shape=(10, 10), chunks=(5, 5))
+        >>> bar[...] = 42
+        >>> store.close()  # don't forget to call this when you're done
+    """
+
+    def __init__(self, path, **kwargs):
+        import sqlite3
+
+        kwargs.setdefault('timeout', 5.0)
+        kwargs.setdefault('cached_statements', 100)
+
+        # normalize path
+        if path != ':memory:':
+            path = os.path.abspath(path)
+
+        # store properties
+        self.path = path
+        self.kwargs = kwargs
+
+        # open database
+        self.db = sqlite3.connect(
+            self.path,
+            detect_types=0,
+            isolation_level=None,
+            check_same_thread=False,
+            **self.kwargs
+        )
+
+        # handle keys as `str`s
+        self.db.text_factory = str
+
+        # get a cursor to read/write to the database
+        self.cursor = self.db.cursor()
+
+        # initialize database with our table if missing
+        self.cursor.execute(
+            'CREATE TABLE IF NOT EXISTS zarr(k TEXT PRIMARY KEY, v BLOB)'
+        )
+
+    def __getstate__(self):
+        return self.path, self.kwargs
+
+    def __setstate__(self, state):
+        path, kwargs = state
+        self.__init__(path=path, **kwargs)
+
+    def close(self):
+        """Closes the underlying database."""
+
+        # close cursor and db objects
+        self.cursor.close()
+        self.db.close()
+
+    def __getitem__(self, key):
+        value = self.cursor.execute('SELECT v FROM zarr WHERE (k = ?)', (key,))
+        for v, in value:
+            return v
+        else:
```

Comment:
Is the `else:` clause necessary here? I'm not sure this is the intended use of `for/else` syntax.

Reply:
Yes, this amounts to handling a missing key.

`sqlite3`'s `execute` always returns an iterator yielding `tuple`s (even if there is only one value). It seemed easy enough to unpack that value with the `for`-loop and `return` that value.

AFAIK (and the tests seem to confirm this) the `else` is correctly handling the case when SQLite doesn't find a matching key. If there was a match, we would have `return`ed and this couldn't have been reached. This amounts to the iterator, `value`, being empty.

Suppose we could do it a different way though.

Reply:
This is the first time I've ever encountered for/else syntax in Python, but [reading around](https://docs.python.org/3/tutorial/controlflow.html#break-and-continue-statements-and-else-clauses-on-loops) it suggest that it doesn't do what you might think, i.e., in fact the else block will get executed even if the iterator in the for block is not empty. The for/else construct is intended for situations where you are using `break` to sometimes break out of the loop. 

In this case it happens to work, because the `return` statement inside the loop means that the `else` block is only visited if the iterator is empty, but that is a happy accident.

In other words, I think it would be simpler and avoid confusion if it was just:

```
for v, in value:
    return v
raise KeyError(k)
```

Reply:
Happy to drop the `else` if you prefer. Have pushed such a change.